Add support for specifying custom device pixel ratio.

1chandu opened the pull request:
<!-- For feature, feature enhancement or bug fix, create an issue first and finish To Do List there -->
<!-- Anything doesn't work as expected is a bug, including code, doc and test -->
For #1154 
<!-- For other PRs without open issue -->
#### Background
- Add support to specify custom device pixel ratio, to enable applications to perform anti aliasing through down sampling. Also helps users to improve performance by setting ratio to a value less than 1
- Group all related code into a single file, add unit tests.
- This is for advanced use case, hence overloading `useDevicePixels` prop, also not documenting intentionally, in future we can make it public API.
- Notes
   - When device pixel ration is high, [depending on implementation](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/drawingBufferWidth), drawing buffer size gets clamped and may not match with canvas size, hence when mapping from window to device pixels, we need to use allocated drawBufferSize, instead of what pixel ratio we used.
   - When drawing buffer size is clamped, it may or may not maintain aspect ratio, hence we need x and y ratios separately to exactly map window to device pixels (like for picking cases).
<!-- For all the PRs -->
#### Change List
- Fix typos
- Add support for custom device pixel ratio

ibgreen commented on modules/core/src/lib/animation-loop-proxy.js:
```diff
@@ -243,9 +247,10 @@ export default class AnimationLoopProxy {
-      const devicePixelRatio = this.useDevicePixels ? window.devicePixelRatio || 1 : 1;
-      const width = this.canvas.clientWidth * devicePixelRatio;
-      const height = this.canvas.clientHeight * devicePixelRatio;
+      // TODO: get access to gl context and use 'mapToDevicePosition'
```
Why do we not have access to the gl context here?

1chandu replied:
Not sure, but yes we might be able to , I will work with @xiaoji on this.

1chandu replied:
Added a TODO , can be addressed in a different PR.

ibgreen commented on modules/core/src/lib/animation-loop-proxy.js:
```diff
@@ -243,9 +247,10 @@ export default class AnimationLoopProxy {
-      const width = this.canvas.clientWidth * devicePixelRatio;
-      const height = this.canvas.clientHeight * devicePixelRatio;
+      // TODO: get access to gl context and use 'mapToDevicePosition'
+      const devicePixelRatio = getDevicePixelRatio(this.useDevicePixels, window.devicePixelRatio);
```
What do we gain by passing in the value of the global `window.devicePixelRatio` in every invocation to `getDevicePixelRatio`, as opposed to referencing it inside `getDevicePixelRatio`?

1chandu replied:
Intention was to de-couple usages `window` inside this utility method, so this can be used in non window (node) environments, also we already have reduced accessing `window.devicePixelRatio` to very few locations in the code base.

ibgreen replied:
> Intention was to de-couple usages window inside this utility method, so this can be used in non window (node) environments, 

In that case, it would still seem helpful to move the `window` reference inside `getDevicePixelRatio`, so that we only need to check for `typeof window === 'undefined'` in one place...

1chandu replied:
Done.

ibgreen commented:
Rereading your notes:
> This is for advanced use case, hence overloading useDevicePixels prop, also not documenting intentionally, in future we can make it public API.

- I think it is worth doing a thorough effort here, as this type of problems tends to keep coming back, and patching on top of old partial fixes is not great.
- I have no issues with it being experimental (underscore) so that we can change it, but I think we should document it just like we do in e.g. deck.gl (assuming we make an extra effort to audit this etc).

> When device pixel ration is high, depending on implementation, drawing buffer size gets clamped and may not match with canvas size, hence when mapping from window to device pixels, we need to use allocated drawBufferSize, instead of what pixel ratio we used.
> When drawing buffer size is clamped, it may or may not maintain aspect ratio, hence we need x and y ratios separately to exactly map window to device pixels (like for picking cases).

I can understand that we would need to check for this case, but why would we need to so separately for `x` and `y`?

1chandu commented:
> I can understand that we would need to check for this case, but why would we need to so separately for `x` and `y`?

`drawingBufferWidth` and `drawingBufferHeight` are some times clamped so that aspect ratio is not maintained compared to `width` and `height` values of the canvas that have been set, and using just one ratio, will lead to invalid CSS to Device pixel conversion. Hence we need to maintain both.

1chandu commented:
@ibgreen @Pessimistress @tsherif

I should have also made it clear, as discussed earlier, this was proposed quick interim solution, until we have `effects` based anti-aliasing support. And I want to get the team's opinion if we still want to do this, given how `drawingBufferWidth` and `drawingBufferHeight` are calculated (`1. can be clamped down to implementation limit and 2. Need to maintain two ratio given aspect ratio is not maintained when device pixel ratio is high) and resulting code changes.  I am more worried about having two ratios and corresponding changes to deck.gl's project module (uniforms and shader code). Do able, but do we have strong enough use case. I can take pass on what deck.gl changes needed for this.

ibgreen commented:
> . And I want to get the team's opinion if we still want to do this, 

Understand that this was proposed as a quick fix and that things get more complicated when addressing the review comments will cause "scope creep".

You'll have to make the call for whether to abort, land or extend this PR. I personally think you have the bones of a great solution here and would not abort at this stage.

If it would be helpful with more input, I'd be happy to discuss the open questions related to my previous comments. Just as long as we don't get "two chefs in the kitchen". E.g. if @tsherif provides directions, just go with those.

tsherif commented:
Generally looks good. I think we should change the name from `devicePixelRatio` since it's no longer the device ratio if it's customized (if we're worried about breaking, this could just be a TODO note for now).

The aspect ratio should be maintained even if clamped. This simply means clamping the larger dimension and calculating the smaller one based on aspect ratio. You can clamp based on `gl.getParamter(gl.MAX_TEXTURE_SIZE)`.

1chandu commented:
Some additional data:
System used Macbook Pro with ATI Radeon Pro 560X GPU.
Max Texture Size: 16384

window size:  800 x 600
green: aspect ratio is maintained, red: aspect ratio not maintained
```diff

+ When `useDevicePixels` set to 5,  canvas width and height set to 4000 X 3000: drawingBufferWidth X Height = 4000 X 3000, xRatio and yRatio are equal to 5.

- When `useDevicePixels` set to 6, canvas width and height set to 4800 X 3600, computed drawingBufferWidth X Height = 4729 X 3547 , xRatio: 5.91125 and yRatio: 5.911666666666667
```

ibgreen commented:
`GL.MAX_RENDERBUFFER_SIZE` on my machine appears to be 	`16384` (see the Framebuffer box in ttps://webglreport.com/?v=2), so you are probably not hitting a limit on each side, rather seems like you are hitting a total allocation limit of 16 Mega "pixels"?

(4729 * 3547 = 16773763, which is just below 2^^24 = 16777216.)

Not sure we can predict this limit accurately (it may depend on available GPU memory).

One strategy is to try to allocate and back off if we can't get what we want?

1chandu commented on modules/webgl/src/utils/device-pixels.js:
```diff
@@ -0,0 +1,43 @@
+}
+
+// Maps CSS pixel position to device pixel position
+export function cssToDevicePixels(gl, cssPixel, yInvert = true) {
```
this will be used in deck.gl use cases.

1chandu commented:
> Looks fine to me. I like the well named functions, but I am a little surprised to see so many new lines of code and comparatively few lines removed. +40 lines net seems like a fairly high cost for this feature.

Assuming you are referring to 40 lines added in `device-pixels.js`, some custom pixel conversion code is removed from examples, for completeness  I also added Device to CSS utility method, which is not used anywhere.

> Have we found all the places where canvas widths are accessed and made sure the new helper methods are used?

yes, every were we are accessing devicePixelRatio to perform conversion is replaced with these utility methods. Also this shouldn't not have any impact when `useDevicePixels` is used as boolean, only when custom ratio is used, updated docs to clarify this.

I still need to do that in deck.gl, deck.gl is accessing devicePixelRatio from `window` object that needs to be updated, I will open deck.gl PR next.

tsherif commented:
I don't like adding 3 new functions to the public API surface for what should be a lightweight experimental feature. `cssToDevicePixels` should be removed if it's not used anywhere, and these functions don't really fit in the WebGL module. Why not just have a property `AnimationLoop.pixelRatio` that the application can refer to to make make whatever calculations it needs?

ibgreen commented:
Not sure this helps, just some more grist for the mill:

- The integration between canvas and context is specified in WebGL so it is not completely out of scope for the webgl module. See for instance https://webgl2fundamentals.org/webgl/lessons/webgl-resizing-the-canvas.html. This integration has also been surprisingly hard to get right in all cases, so not unreasonable to provide some utils.

So maybe these functions can fit well with the part of our webgl lib that creates and destroys contexts (the `gl.canvas` part). Someone using luma to build a framework would arguably use them. I think you said deck will also be using them?

Whether they should be exported as public or kept internal, and whether we should include functions that are not currently used, could depend on the use case. I think this PR was initiated in response to a user request, so we could evaluate against that need. 

If we then see that the new `CSStoDevice` function is useful to that use case, and we can foresee that the user would immediately ask us to add the `DeviceToCSS` function (perhaps to let him handle events), then maybe keep it.

Another consideration is that for useful things that are rarely used, we also have luma.gl/addons to keep them out of the core. But in this case, one of the functions is used heavily in the core so it would be weird to have the inverse function in another module.

tsherif commented:
Couldn't the current pixel ratio just be calculated once as `[drawingBufferWidth / clientWidth, drawingBufferHeight / clientHeight]` (updated on resize, etc.) and then multiplying or dividing by those numbers would be enough to replace all the helper functions?

Pessimistress commented:
> Couldn't the current pixel ratio just be calculated once as [drawingBufferWidth / clientWidth, drawingBufferHeight / clientHeight] (updated on resize, etc.) 

It's the other way around - we set drawing buffer dimensions based on the user defined pixel ratio.

1chandu commented:
@ibgreen @tsherif 

I see following two reasons to provide `cssToDevicePixels` and `cssToDeviceRatio` (will add it) utility methods :

#1. Yes, this PR was initiated due to a user request for custom pixel ratio, but during the process, I did discover the problem, where when a pixel ratio is too high, final drawingBuffer width height getting clamped resulting invalid aspect ratio, due to this fact neither `window.devicePixelRatio` nor user provided `useDevicePixelRatio` contain actual ratio that is being used, instead it is the `drawingBufferWidth/clientWidth`.  Given these two are the props of `gl`, these should be part of `webgl`

#2. Several places in `deck.gl`, we are repeating the calculations (what `cssToDevicePixels` does) and these should be provided as utility methods to avoid code redundancy.

I am ok to keep these in a different module instead of `webgl`.

1chandu commented:
I am working to fix some test failures but for context these are deck.gl changes : https://github.com/uber/deck.gl/commit/65b1bc37eeba0f20b4db42b4cac2864853629abd

tsherif commented:
> It's the other way around - we set drawing buffer dimensions based on the user defined pixel ratio.

I understand that, but once you have the input from the user, you can calculate the ratio once and maintain it as a data property.

tsherif commented:
> I see following two reasons to provide cssToDevicePixels and cssToDeviceRatio (will add it) utility methods

Since we're not sure what we want the final API to look like (e.g. the `*Device*` names seem inappropriate since we're not talking about the device ratio anymore), I'd prefer we keep the changes as small as possible. Two possible minimal approaches:
- maintain `AnimationLoop.pixelRatio`, a vec2 that contains the current [drawingBufferWidth/clientWidth, ...]
- one function `getPixelRatio(gl)` that returns a vec2 similar to the above

1chandu commented:
@tsherif as discussed, removed deviceToCSS* utility methods (just keeping CssToDevice*) that are not being used any where, also 
- Made `getDevicePixelRatio` private to `webgl` module, every one else should be using `deviceToCSS*` utility methods for conversion.
- Added code to handle a case where `clientWidth` and `clientHeight` are 0 , basically matched behavior to old code. (It is happening while running deck.gl unit tests).

Pessimistress commented on modules/webgl/src/utils/device-pixels.js:
```diff
@@ -0,0 +1,42 @@
+  // since we are rounding to nearest, when dpr > 1, edge pixels may point to out of bounds value, clamp to the limit
+  const x = Math.min(Math.round(pixel[0] * ratio), width - 1);
+  const y = yInvert
+    ? Math.max(0, height - Math.ceil(ratio) - Math.round(pixel[1] * ratio))
```
Why `Math.ceil(ratio)`? Is it not possible to get the bottom pixel if ratio = 2?

1chandu replied:
`ceil` is only used for y invert cases, when pixel ratio is fraction, without `ceil` converted pixel location is not an integer. I have unit tests covering this case.

When pixel ratio is `2` , you can still select the bottom pixel, css 0 points to device 0, css 1 points to device 2 etc.

1chandu commented on docs/api-reference/webgl/device-pixels.md:
```diff
@@ -0,0 +1,40 @@
+
+Depending on `useDevicePixels` value, returns current device pixel ratio or 1.
+
+* `useDevicePixels` (Boolean) - whether to use device resolution or not.
```
This will be private method, removed it here, I have section in `animation-loop.md` explaining the experimental API (`Number`).

1chandu commented on modules/webgl/src/utils/device-pixels.js:
```diff
@@ -0,0 +1,43 @@
+export function getDevicePixelRatio(useDevicePixels) {
+  const windowRatio = typeof window === 'undefined' ? 1 : window.devicePixelRatio || 1;
+  if (Number.isFinite(useDevicePixels)) {
+    return useDevicePixels <= 0 ? windowRatio : useDevicePixels;
```
Done.

1chandu commented:
@Pessimistress , updated `cssToDevicePixels` to return a range of valid pixels :
`{ low: [x, y], high: [x, y]}`
When DPR <=1, `low` and `high` will be identical, when DPR > 1, `low` and `high` are not equal and represent entire range of device pixel positions. 

On the naming, other options considered: 1. `{begin: [x, y], end: [x, y]}` , 2. `[[x, y], [x, y]]`.

Will update the docs before merging.

Pessimistress commented on docs/api-reference/webgl/device-pixels.md:
```diff
@@ -0,0 +1,35 @@
+* `cssPixels` (Array) - Array in [x, y] form, where x and y are location in CSS window.
+* `yInvert` (Boolean, optional, default: true) - when true it will perform y-inversion when converting to Device pixels.
+
+Returns Device pixel location, [x, y].
```
Update?

Pessimistress commented on modules/webgl/src/utils/device-pixels.js:
```diff
@@ -0,0 +1,66 @@
+  // If next pixel's position is clamped to boundary, use it as is, otherwise subtract delta for current pixel boundary
+  const yHigh = t === yBoundary ? t : t - yDelta;
+
+  return {
```
Use `{x, y, width, height}` consistent with frameworks conventions and the [DOMRect](https://developer.mozilla.org/en-US/docs/Web/API/DOMRect) object?

Pessimistress commented on modules/webgl/src/context/context.js:
```diff
@@ -203,3 +196,30 @@ function getVersion(gl) {
+  do {
+    const {clientWidth, clientHeight} = gl.canvas;
+    gl.canvas.width = Math.ceil(clientWidth * devicePixelRatio);
+    gl.canvas.height = Math.ceil(clientHeight * devicePixelRatio);
```
This no longer works if `clientWidth` is not valid?

1chandu replied:
I am using `clientSizeValid` variable , if `clientWidth` or `clientHeight` is not valid, we exit the loop and the behavior remains same.

Pessimistress replied:
Event if `clientSizeValid` is `false` you have already overwritten the canvas width and height.

1chandu replied:
we were doing the same before these changes, I can set a safe value , may be `1` if width or height not valid ?

1chandu replied:
Added safe back up value of 1.

Pessimistress commented on modules/webgl/src/context/context.js:
```diff
@@ -166,15 +166,8 @@ export function destroyGLContext(gl) {
-    gl.canvas.width = width * devicePixelRatio;
-    gl.canvas.height = height * devicePixelRatio;
-
+    const devicePixelRatio = getDevicePixelRatio(options.useDevicePixels);
```
Looks like a breaking change - `options.width` and `options.height` are no longer supported.

1chandu replied:
`options.width` and `options.height` are not used by any one who is calling `resizeGLContext`, also these are not documented.
Are there any use cases I am missing? Do you think I should add this back?

Pessimistress replied:
Let's be safe and add it back for now. You can make a comment to revisit the API in v8.

1chandu replied:
Added.

Pessimistress commented on modules/webgl/src/utils/device-pixels.js:
```diff
@@ -0,0 +1,66 @@
+  const yBoundary = yInvert ? 0 : height - 1;
+  const yDelta = yInvert ? -1 : 1;
+  // If next pixel's position is clamped to boundary, use it as is, otherwise subtract delta for current pixel boundary
+  const yHigh = t === yBoundary ? t : t - yDelta;
```
`yLow` should always be smaller than `yHigh`.

1chandu replied:
As you mentioned below, I will use `{x, y, width, height}`, where `x` and `y` are smaller of the rect bounds.

Pessimistress commented on modules/webgl/src/context/context.js:
```diff
@@ -203,3 +196,34 @@ function getVersion(gl) {
+  let aspectRatioValid = false;
+
+  // NOTE: if options.width and options.height not used remove in v8
+  let clientWidth = 'width' in options ? options.width : gl.canvas.clientWidth || gl.canvas.width;
```
Just add `|| 1`?

1chandu replied:
Done. (given in handles `0` too)
